internal/response-consumer/handler.go

```diff
@@ -164,6 +156,100 @@ func (this *handler) onMessage(ctx context.Context, msg *k.Message) {
 	}
 }
 
+func buildCaseExpr(caseOptions []string) clause.Expr {
+	argsLen := len(caseOptions)
+	caseExprStr := "CASE "
+	varsArray := make([]interface{}, argsLen)
+
+	for i, option := range caseOptions {
+		varsArray[i] = clause.Expr{SQL: option}
+
+		if i == argsLen-1 {
+			caseExprStr += fmt.Sprintf("ELSE (?) END")
+		} else if i%2 == 0 {
+			caseExprStr += fmt.Sprintf("WHEN (?) ")
+		} else {
+			caseExprStr += fmt.Sprintf("THEN (?) ")
+		}
+	}
+
+	return clause.Expr{
+		SQL:  caseExprStr,
+		Vars: varsArray,
+	}
+
+}
+
+func satAssignmentWithCase(responseFull bool, updateHost db.RunHost) map[string]interface{} {
+	satSequence, status, log := *updateHost.SatSequence, updateHost.Status, updateHost.Log
+
+	satNullClause := "sat_sequence IS NULL"
+
+	statusCase := []string{
+		fmt.Sprintf("%s OR sat_sequence < %d", satNullClause, satSequence),
+		fmt.Sprintf("'%s'", status),
+		"status",
+	}
+
+	satSeqCase := []string{
+		fmt.Sprintf("%s OR sat_sequence < %d", satNullClause, satSequence),
+		strconv.Itoa(satSequence),
+		"sat_sequence",
+	}
+
+	logCase := []string{
+		fmt.Sprintf("%s OR sat_sequence < %d", satNullClause, satSequence),
+		fmt.Sprintf("'%s'", log),
+		"log",
+	}
+
+	if !responseFull {
+		logCase = []string{
+			fmt.Sprintf("(%s AND %d > 0) OR sat_sequence + 1 < %d", satNullClause, satSequence, satSequence),
+			fmt.Sprintf("log || '&#8230;' || '%s'", log),
```

Comment:
this looks vulnerable to SQL injection

Reply:
Great catch! I tried simulating a SQL injection by changing the a random console log in the tests and it was in fact vulnerable to SQL injection. I switched it to using variable substitution, and that seems to escape the SQL. Gorm's documentation doesn't say a [whole lot about methods that escapes SQL statements](https://gorm.io/docs/security.html#SQL-injection-Methods). However, `clause.Expr{}` seems like it has [logic built-in to escape SQL statements](https://github.com/go-gorm/gorm/blob/master/clause/expression.go).